xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-war/src/main/webapp/resources/js/xwiki/editors/dataeditors.js

```diff
@@ -17,825 +17,754 @@
+      // We want to the user to be prevented if he tries to leave the editor before saving.
+      $(document).on("beforeunload", function(event) {
+        if (Object.keys(this.editorStatus.addedXObjects).length > 0
+          || Object.keys(this.editorStatus.deletedXObjects).length > 0) {
+          event.preventDefault();
+          event.returnValue = "";
         } else {
-          this.editorStatus.savedXObjects[xclassName].concat(this.editorStatus.addedXObjects[xclassName]);
-          this.editorStatus.savedXObjects[xclassName].sort(this.numberSort);
+          return;
         }
-      }
-      this.editorStatus.addedXObjects = {};
-      this.editorStatus.deletedXObjects = {};
-    }.bindAsEventListener(this));
-
-    // in case of cancel we just clean everything so that we don't get any warnings for leaving the page without saving.
-    document.observe('xwiki:actions:cancel', function (event) {
-      // Remove fields about deleted and added objects
-      $$('input[name=deletedObjects]').each(function(item) {
-        item.remove();
       });
-      $$('input[name=addedObjects]').each(function(item) {
-        item.remove();
-      });
-      this.editorStatus.addedXObjects = {};
-      this.editorStatus.deletedXObjects = {};
-    }.bindAsEventListener(this));
+    },
+    /**
+     * Sort function to allow sorting an array of integer.
+     * Using [3, 0, 1].sort(numberSort) produces [0, 1, 3].
+     */
+    numberSort : function (a, b) {
+      return a - b;
+    },
+    /**
+     * Extract the xclass name from the XDOM id of the xclass.
+     */
+    getXClassNameFromXClassId : function (xclassId) {
+      return xclassId.substring("xclass_".length);
+    },
+    /**
+     * Extract the xclass name from the XDOM id of an xobject.
+     */
+    getXClassNameFromXObjectId : function (xobjectId) {
+      return xobjectId.substring("xobject_".length, xobjectId.lastIndexOf('_'));
+    },
+    /**
+     * Extract the xobject number from the XDOM id of an xobject.
+     */
+    getXObjectNumberFromXObjectId : function (xobjectId) {
+      return xobjectId.substring(xobjectId.lastIndexOf('_') + 1);
+    },
+    /**
+     * Returns true if an object exists with the given classname and number.
+     */
+    xObjectAlreadyExist : function (xclassName, objectNumber) {
+      return (this.editorStatus.savedXObjects[xclassName] !== undefined
+        && this.editorStatus.savedXObjects[xclassName].indexOf(objectNumber) !== -1)
+        || (this.editorStatus.addedXObjects[xclassName] !== undefined
+          && this.editorStatus.addedXObjects[xclassName].indexOf(objectNumber) !== -1);
+    },
+    /**
+     * Returns the xobject DOM element for the given class and number.
+     */
+    getXObject : function (xclassName, objectNumber) {
+      let expectedId = 'xobject_' + xclassName + '_' + objectNumber;
+      // We cannot use # + expectedId because of the dots to escape.
```

Comment:
You could use https://developer.mozilla.org/en-US/docs/Web/API/CSS/escape_static . In fact, you probably should use it anyway to be safe.

Reply:
Fixed.